File: machine/machine.go
package machine

import (
	"container/list"
	"fmt"
	"io/ioutil"
	"log"
	"regexp"
	"strings"
)

type void struct{}
type changes struct {
	symbol    rune
	direction bool
	state     string
}

// Machine defines the basic data structure for a Turing machine
type Machine struct {
	symbols     map[rune]void
	blankSymbol rune
	states      map[string]void
	startState  string
	endStates   map[string]void
	transitions map[string]map[rune]changes

	currentState string
	tape         *list.List
	head         *list.Element
}

// NewMachine returns a Machine instance
func NewMachine() *Machine {
	m := Machine{}
	m.symbols = make(map[rune]void)
	m.states = make(map[string]void)
	m.endStates = make(map[string]void)
	m.transitions = make(map[string]map[rune]changes)

	return &m
}

// SetSymbol stores a symbol in the Machine definition
// The symbol can be set as the blank symbol only if it is being set for the first time
// The program halts if a symbol is set as the blank symbol when the blank symbol is already set
func (m *Machine) SetSymbol(symbol rune, isBlank bool) {
	m.symbols[symbol] = void{}
	if isBlank {
		if m.blankSymbol == 0 {
			m.blankSymbol = symbol
		} else {
			log.Fatalf("Machine got blank symbol '%c' which is already set to '%c'", symbol, m.blankSymbol)
		}
	}
}

// SetState stores a state in the Machine definition
// The state can be set as the start state and/or one of the end states if it is being set for the first time
// The program halts if a state is set as the start state when the start state is already set
func (m *Machine) SetState(state string, isStart, isEnd bool) {
	m.states[state] = void{}
	if isStart {
		if m.startState == "" {
			m.startState = state
		} else {
			log.Fatalf("Machine got start state '%s' which is already set to '%s'", state, m.startState)
		}
	}

	if isEnd {
		m.endStates[state] = void{}
	}
}

// SetTransition stores a transition in the Machine definition
func (m *Machine) SetTransition(currentState string, currentSymbol, nextSymbol rune, direction bool, nextState string) {
	if _, ok := m.transitions[currentState]; !ok {
		m.transitions[currentState] = make(map[rune]changes)
	}

	m.SetState(currentState, false, false)
	m.SetState(nextState, false, false)
	m.SetSymbol(currentSymbol, false)
	m.SetSymbol(nextSymbol, false)

	m.transitions[currentState][currentSymbol] = changes{nextSymbol, direction, nextState}
}

func (m *Machine) Reset(input string) {
	m.currentState = m.startState
	m.tape = list.New()
	if input == "" {
		m.tape.PushBack(m.blankSymbol)
	} else {
		for _, r := range input {
			m.tape.PushBack(r)
		}
	}
	m.head = m.tape.Front()
}

// Step defines one transition by the Turing machine
// and returns whether or not the machine halts due to no valid transition
func (m *Machine) Step() bool {
	change, ok := m.transitions[m.currentState][m.head.Value.(rune)]
	if !ok {
		return true
	}

	m.head.Value = change.symbol

	if change.direction {
		if temp := m.head.Next(); temp != nil {
			m.head = temp
		} else {
			m.tape.PushBack(m.blankSymbol)
			m.head = m.tape.Back()
		}
	} else {
		if temp := m.head.Prev(); temp != nil {
			m.head = temp
		} else {
			m.tape.PushFront(m.blankSymbol)
			m.head = m.tape.Front()
		}
	}

	m.currentState = change.state
	return false
}

// Compute defines a computation by the Turing machine on an input string
// and returns whether the machine halted and whether it accepted the input string
func (m *Machine) Compute(input string, maxSteps uint) (bool, bool) {
	m.Reset(input)
	var steps uint
	halt := false

	for steps < maxSteps && !halt {
		steps++
		halt = m.Step()
	}

	_, accepted := m.endStates[m.currentState]
	return halt, halt && accepted
}

// Parse parses the definition in a specified filename
// and returns the corresponding Machine
func (m *Machine) Parse(filename string) {
	commentRE := regexp.MustCompile(`(#.*)`)

	data, err := ioutil.ReadFile(filename)
	if err != nil {
		log.Fatal(err)
	}

	lines := strings.Split(string(data), "\n")
	for i := range lines {
		lines[i] = commentRE.ReplaceAllString(lines[i], "")
		lines[i] = strings.TrimSpace(lines[i])
	}

	blankSymbolRE := regexp.MustCompile(`^'(.)'$`)
	stateRE := regexp.MustCompile(`^([a-zA-Z_][a-zA-Z0-9_]*)([.*])?$`)
	transitionRE := regexp.MustCompile(`^'(.)'\s+'(.)'\s+([<>])\s+([a-zA-Z_][a-zA-Z0-9_]*)$`)
	currentState := ""

	for index, l := range lines {
		if l == "" {
			continue
		}

		bMatch := blankSymbolRE.FindStringSubmatch(l)
		if len(bMatch) > 0 {
			m.SetSymbol(rune(bMatch[1][0]), true)
			continue
		}

		sMatch := stateRE.FindStringSubmatch(l)
		if len(sMatch) > 0 {
			currentState = sMatch[1]
			isStart := false
			isEnd := false

			for _, r := range sMatch[2] {
				switch r {
				case '.':
					isEnd = true
				case '*':
					isStart = true
				}
			}

			m.SetState(currentState, isStart, isEnd)
			continue
		}

		tMatch := transitionRE.FindStringSubmatch(l)
		if len(tMatch) > 0 {
			if currentState == "" {
				break
			}

			direction := false
			switch rune(tMatch[3][0]) {
			case '<':
				direction = false
			case '>':
				direction = true
			}

			m.SetTransition(currentState, rune(tMatch[1][0]), rune(tMatch[2][0]), direction, tMatch[4])
			continue
		}

		log.Fatalf("Error at line %d\n", index+1)
	}
}

// Graph returns a definition of the Machine in Graphviz (https://www.graphviz.org/)
func (m *Machine) Graph() string {
	var builder strings.Builder
	builder.WriteString("digraph machine {\n")
	builder.WriteString("\trankdir=LR;\n")
	builder.WriteString("\tsize=\"8,5\";\n")

	builder.WriteString("\n")
	builder.WriteString("\tnode [shape = point]; 0;\n")

	for state := range m.endStates {
		builder.WriteString(fmt.Sprintf("\tnode [shape = doublecircle]; %s;\n", state))
	}

	builder.WriteString("\tnode [shape = circle];\n")
	builder.WriteString(fmt.Sprintf("\t0 -> %s;\n", m.startState))

	for currentState := range m.states {
		for currentSymbol, change := range m.transitions[currentState] {
			dir := 'L'
			if change.direction {
				dir = 'R'
			}

			builder.WriteString(fmt.Sprintf("\t%s -> %s [ label = \"'%c', '%c', '%c'\" ];\n", currentState, change.state, currentSymbol, change.symbol, dir))
		}
	}

	builder.WriteString("}\n")

	return builder.String()
}

File: src/alan/machine.py
import imageio
import re
from pydot import Dot, Edge, Node, Subgraph

class Machine:
    """
    The abstract class for a Turing Machine.
    """

    def __init__(self):
        """
        Initializes the Machine object.
        """
        self._symbols = set()
        self._blank_symbol = None
        self._states = set()
        self._start_state = None
        self._end_states = set()
        self._transitions = {}

        self._current_state = None
        self._tape = None
        self._head = None


    def _set_symbol(self, symbol, blank=False):
        """
        Adds a symbol to the set of symbols of the Machine.
        Sets the blank symbol only if it is not already set.

        Arguments:
            symbol {str} -- symbol

        Keyword Arguments:
            blank {bool} -- whether the symbol must be set as the blank symbol (default: {False})
        """
        self._symbols.add(symbol)

        try:
            assert self._blank_symbol == None or not blank
            if blank:
                self._blank_symbol = symbol
        except:
            raise Exception(f"Machine got blank symbol '{symbol}' which is already set to '{self._blank_symbol}'")


    def _set_state(self, state, start=False, end=False):
        """
        Adds a state to the set of states of the Machine.
        Sets the start state only if it is not already set.

        Arguments:
            state {str} -- state

        Keyword Arguments:
            start {bool} -- whether the symbol must be set as the start state (default: {False})
            end {bool} -- whether the symbol must be set as the end state (default: {False})
        """
        self._states.add(state)
        if end:
            self._end_states.add(state)

        try:
            assert self._start_state == None or not start
            if start:
                self._start_state = state
        except:
            raise Exception(f"Machine got start state '{state}' which is already set to '{self._start_state}'")


    def _set_transition(self, current_state, current_symbol, next_symbol, direction, next_state):
        """
        Adds a transition to the set of transitions of the Machine.

        Arguments:
            current_state {str} -- current State
            current_symbol {str} -- current Symbol
            next_symbol {str} -- next Symbol
            direction {bool} -- direction
            next_state {str} -- next State
        """
        self._set_symbol(current_symbol)
        self._set_symbol(next_symbol)
        self._set_state(current_state)
        self._set_state(next_state)

        if self._transitions.get(current_state) is None:
            self._transitions[current_state] = {}

        self._transitions[current_state][current_symbol] = (next_symbol, direction, next_state)


    def parse(self, definition):
        """
        Parses a definition for a Machine object.

        Arguments:
            definition {str} -- definition
        """
        comment_re = re.compile(r"(#.*)")
        state_re = re.compile(r"^([a-zA-Z_][a-zA-Z0-9_]*)([.*])?$")
        transition_re = re.compile(r"^'(.)'\s+'(.)'\s+([<>])\s+([a-zA-Z_][a-zA-Z0-9_]*)$")

        lines = list(filter(lambda x : len(x) > 0, map(lambda x: comment_re.sub("", x).strip(), definition.split("\n"))))
        self._set_symbol(lines.pop(0)[1], True)

        scope_state = None
        for l in lines:
            try:
                g = state_re.match(l).groups()
                scope_state = g[0]
                special = g[1] if g[1] is not None else ""
                self._set_state(scope_state, "*" in special, "." in special)
            except:
                pass

            try:
                g = transition_re.match(l).groups()
                self._set_transition(scope_state, g[0], g[1], g[2] is ">", g[3])
            except:
                pass


    def reset(self, tape):
        """
        Resets the tape, head and state of the Machine.

        Arguments:
            tape {[type]} -- string of tape symbols
        """
        self._current_state = self._start_state
        self._tape = list(tape)
        self._head = 0


    def step(self):
        """
        Performs one Machine step.

        Returns:
            {bool} -- whether or not the Machine halted
        """
        try:
            current_symbol = self._tape[self._head]
            next_symbol, direction, self._current_state = self._transitions.get(self._current_state).get(current_symbol)
        except:
            return True

        self._tape[self._head] = next_symbol
        self._head += 1 if direction else -1

        if self._head < 0:
            self._tape.insert(0, self._blank_symbol)
            self._head = 0
        elif self._head >= len(self._tape):
            self._tape.append(self._blank_symbol)
            self._head = len(self._tape) - 1

        return False


    def run(self, tape, max_steps=200, animate=False, **kwargs):
        """
        Performs a computation by the Machine on a tape

        Arguments:
            tape {str} -- string of tape symbols

        Keyword Arguments:
            max_steps {int} -- maximum steps before forcing halt (default: {200})
            animate {bool} -- whether to call animation functionality (default: {False})

        Returns:
            ({bool}, {bool}, {str}) -- whether it halted, whether it accepted, final tape
        """
        self.reset(tape)
        halt = False

        if animate:
            try:
                assert kwargs.get("filename") is not None
            except:
                raise Exception("Specify a filename to save the animation")
            images = []


        for _ in range(max_steps):
            if animate:
                images.append(imageio.imread(self.graph(False).create(prog="dot", format="png")))
                images.append(imageio.imread(self.graph(True).create(prog="dot", format="png")))

            halt = self.step()
            if halt:
                break

        if animate:
            imageio.mimsave(kwargs.get("filename"), images, fps=kwargs.get("fps"))

        return halt, halt and self._current_state in self._end_states, "".join(self._tape).strip(self._blank_symbol)


    def graph(self, context=None, **kwargs):
        """
        Returns a graph object

        Keyword Arguments:
            context {bool|None} -- controls the animations (default: {None})

        Returns:
            [pydot.Dot] -- graph object
        """
        graph = Dot(graph_type="digraph", rankdir=("LR" if context is None else "TB"))
        machine_graph =  Subgraph(graph_name="cluster_machine", graph_type="digraph", label="MACHINE")

        for current_state in sorted(self._states):
            node_args = {}
            shape = "circle"
            shape = ("double" if current_state in self._end_states else "") + shape
            node_args["shape"] = shape

            if context is not None and current_state == self._current_state:
                node_args["fillcolor"] = "cyan"
                node_args["style"] = "filled"

            machine_graph.add_node(Node(current_state, **node_args))

        machine_graph.add_node(Node("0", shape="point"))
        machine_graph.add_edge(Edge("0", self._start_state))

        for current_state in self._states:
            transitions = self._transitions.get(current_state)
            if transitions:
                for current_symbol in transitions:
                    next_symbol, direction, next_state = transitions.get(current_symbol)
                    label = f"'{current_symbol}' '{next_symbol}' {'R' if direction else 'L'}"

                    edge_args = {}
                    if context and current_state == self._current_state and current_symbol == self._tape[self._head]:
                        edge_args["color"] = "cyan"
                    machine_graph.add_edge(Edge(current_state, next_state, label=label, **edge_args))

        graph.add_subgraph(machine_graph)
        if context is not None:
            tape_graph = Subgraph(graph_name="cluster_tape", graph_type="digraph", label="TAPE")
            tape = []
            for index in range(-4 + self._head, 5 + self._head):
                tape.append(f"<t{index}> {self._tape[index] if 0 <= index < len(self._tape) else self._blank_symbol}")

            tape_graph.add_node(Node("tape", label="|".join(tape), shape="record"))
            tape_graph.add_node(Node("t0", shape="point"))
            tape_graph.add_edge(Edge("t0", f"tape:t{self._head}"))
            graph.add_subgraph(tape_graph)

        if kwargs.get("filename"):
            graph.write(kwargs.get("filename"), format="png")
            return f"Graph saved to {kwargs.get('filename')}"

        return graph
